File: 21. Merge Two Sorted Lists.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def mergeTwoLists(list1, list2):
    if not list1 and not list2:
        return None

    def get_smallest(l1, l2):
        if not l1 and not l2:
            return None
        elif not l1:
            return l2
        elif not l2:
            return l1
        return min(l1, l2, key=lambda x: x.val)

    head = get_smallest(list1, list2)
    if head == list1:
        list1 = list1.next
    else:
        list2 = list2.next
    cur = head
    while True:
        small = get_smallest(list1, list2)
        if small is None:
            return head
        if small == list1:
            cur.next = list1
            cur = cur.next
            list1 = list1.next
        else:
            cur.next = list2
            cur = cur.next
            list2 = list2.next

File: 21. Merge Two Sorted Lists.c
// https://leetcode.com/problems/merge-two-sorted-lists/description/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode *mk_node(int val, struct ListNode *nxt) {
    struct ListNode *res = malloc(sizeof(struct ListNode));
    res->val = val;
    res->next = nxt;
    return res;
}

struct ListNode* mergeTwoLists(struct ListNode* list1, struct ListNode* list2) {
    struct ListNode *filler = mk_node(-1, NULL);  // temporary, will be removed at the end

    struct ListNode *cur = filler;
    while (list1 || list2) {
        if (!list1) {
            cur->next = mk_node(list2->val, NULL);
            list2 = list2->next;
        } else if (!list2) {
            cur->next = mk_node(list1->val, NULL);
            list1 = list1->next;
        } else if (list1->val <= list2->val) {
            cur->next = mk_node(list1->val, NULL);
            list1 = list1->next;
        } else {
            cur->next = mk_node(list2->val, NULL);
            list2 = list2->next;
        }
        cur = cur->next;
    }
    struct ListNode *ret = filler->next;
    free(filler);
    return ret;
}
